cmd/maya-apiserver/spc-watcher/storagepool_create.go

```diff
@@ -0,0 +1,180 @@
+	return nil
+}
+
+// newCasPool will return a CasPool object
+func newCasPool(spcGot *apis.StoragePoolClaim, reSync bool, pendingPoolCount int) (error, *apis.CasPool) {
+	// Validations for poolType
+	poolType := spcGot.Spec.PoolSpec.PoolType
+	if poolType == "" {
+		return errors.New("aborting storagepool create operation as no poolType is specified"), nil
+	}
+
+	if !(poolType == string(v1alpha1.PoolTypeStripedCPK) || poolType == string(v1alpha1.PoolTypeMirroredCPK)) {
+		return fmt.Errorf("aborting storagepool create operation as specified poolType is %s which is invalid", poolType), nil
+	}
+
+	diskType := spcGot.Spec.Type
+	if !(diskType == string(v1alpha1.TypeSparseCPK) || diskType == string(v1alpha1.TypeDiskCPK)) {
+		return fmt.Errorf("aborting storagepool create operation as specified type is %s which is invalid", diskType), nil
+	}
+	// The name of cas template should be provided as annotation in storagepoolclaim yaml
+	// so that it can be used.
+
+	// Check for cas template
+	casTemplateName := spcGot.Annotations[string(v1alpha1.SPCreateCASTemplateCK)]
+	if casTemplateName == "" {
+		return errors.New("aborting storagepool create operation as no cas template is specified"), nil
+	}
+	// Create an empty CasPool object and fill storagepoolcalim details
+	pool := &v1alpha1.CasPool{}
+	pool.StoragePoolClaim = spcGot.Name
+	pool.CasCreateTemplate = casTemplateName
+	pool.PoolType = spcGot.Spec.PoolSpec.PoolType
+	pool.MinPools = spcGot.Spec.MinPools
+	pool.MaxPools = spcGot.Spec.MaxPools
+	pool.Type = spcGot.Spec.Type
+	pool.ReSync = reSync
+	pool.PendingPoolCount = pendingPoolCount
+
+	// Fill the object with the disks list
+	pool.DiskList = spcGot.Spec.Disks.DiskList
+	// Check for disks
+	diskList := spcGot.Spec.Disks.DiskList
+	// If no disk are specified pool will be provisioned dynamically
+	if len(diskList) == 0 {
+		// newDisksList is the list of disks over which pool will be provisioned
+		err, newDisksList := getCasPoolDisk(pool)
+		if err != nil {
+			return err, nil
+		}
+		// Fill the object with the new disks list
+		pool.DiskList = newDisksList
+	}
+	return nil, pool
+}
+
+// getCasPoolDisk is a wrapper that will call getDiskList function to get the disk lists
+// that will be used to provision a storagepool dynamically
+
+func getCasPoolDisk(cp *apis.CasPool) (error, []string) {
+	// Performing valdations against CasPool fields
+	if cp.MaxPools <= 0 {
+		return fmt.Errorf("aborting storagepool create operation as no maxPool field is specified"), nil
+	}
+	// if no minimum pools were specified it will default to 1.
+	if cp.MinPools <= 0 {
+		glog.Warning("invalid or 0 min pool specified, defaulting to 1")
+		cp.MinPools = 1
+	}
+	if cp.MaxPools < cp.MinPools {
+		return fmt.Errorf("aborting storagepool create operation as maxPool cannot be less than minPool"), nil
+	}
+	// If it is a resync event, MaxPool is the pending pool to be provisioned
```

Comment:
The above 3 validations should be a part of newCasPool function.

Reply:
We discussed this part.
We are not validating it as part of newCasPool function as it is not relevant to manual pool provisioning.
Once we form reconcile logic for manual provisioned pool, this will be moved.

Reply:
If moving the validations to newCasPool() does not cause any regression, I'd recommend to do that in order to maintain consistency. If there is a chance of any regression, we can track with a different backlog. Having the validations at one place for user-specified inputs is better and need not necessarily tie with reconcile requirement.

Reply:
It can cause if users skip maxPool field.
I feel updating this part when reconciliation for manual provisioned pool is done

Reply:
I will track it as a backlog and address it in subsequent PR